File: src/Creational/Builder/InsurancePolicyBuilder.cs
namespace DesignPatterns.Creational.Builder;

public class InsurancePolicyBuilder : IInsurancePolicyBuilder
{
    private readonly InsurancePolicy _policy = new();

    public IInsurancePolicyBuilder WithPolicyType(string policyType)
    {
        _policy.PolicyType = policyType;
        return this;
    }

    public IInsurancePolicyBuilder WithPremium(double premium)
    {
        _policy.Premium = premium;
        return this;
    }

    public IInsurancePolicyBuilder WithCoverAmount(double coverAmount)
    {
        _policy.CoverAmount = coverAmount;
        return this;
    }

    public InsurancePolicy Build()
    {
        return _policy;
    }
}

File: ts/src/creational/builder/InsurancePolicyBuilder.ts
import { InsurancePolicy } from './InsurancePolicy';
import { IInsurancePolicyBuilder } from './IInsurancePolicyBuilder';

export class InsurancePolicyBuilder implements IInsurancePolicyBuilder {
    private policy: InsurancePolicy = new InsurancePolicy();

    withPolicyType(policyType: string): this {
        this.policy.policyType = policyType;
        return this;
    }

    withPremium(premium: number): this {
        this.policy.premium = premium;
        return this;
    }

    withCoverAmount(amount: number): this {
        this.policy.coverAmount = amount;
        return this;
    }

    build(): InsurancePolicy {
        return this.policy;
    }
}
